src/ChatHistoryItem.tsx

```diff
@@ -32,29 +43,28 @@ export const ChatHistoryItem = memo(
       e.preventDefault();
       e.stopPropagation();
       if (editingTitle) {
+        setEditInputValue(optimisticTitle);
         setEditingTitle(false);
-        setEditInputValue(title);
       }
       setIsMenuOpen((prev) => !prev);
     };
 
-    const handleSaveTitle = useCallback(
-      (currentTitle: string) => {
-        const trimmedTitle = currentTitle.trim();
-        if (trimmedTitle === '') {
-          setEditInputValue(title);
-          setEditingTitle(false);
-          return;
-        }
+    const editInputRef = useRef(editInputValue);
 
-        if (trimmedTitle !== title) {
-          updateConversationTitle(id, trimmedTitle);
-        }
+    const handleSaveTitle = useCallback(() => {
+      const currentTitle = editInputRef.current.trim();
 
+      if (currentTitle === '' || currentTitle === optimisticTitle) {
+        setEditInputValue(optimisticTitle);
         setEditingTitle(false);
-      },
-      [title, updateConversationTitle, id],
-    );
+        return;
+      }
+
+      setOptimisticTitle(currentTitle);
+      setEditingTitle(false);
+
+      updateConversationTitle(id, currentTitle);
```

Comment:
- Updates the UI 
  - new title
  - closes edit mode
- Makes db update (and resolves any diff)